mlir/lib/Dialect/Bufferization/IR/BufferizationOps.cpp

```diff
@@ -245,7 +245,8 @@ AllocTensorOp::getBufferType(Value value, const BufferizationOptions &options,
     return getOperation()->emitError("could not infer memory space");
   }
 
-  return getMemRefTypeWithStaticIdentityLayout(getType(), memorySpace);
+  return cast<BufferLikeType>(
```

Comment:
note: unfortunately, i didn't seem to manage to avoid this even with an addition of an implicit ctor: `BufferLikeType(BaseMemRefType)`. I think this is due to also a FailureOr<> wrapper. I wonder yet again whether it makes sense to push a patch to extend FailureOr<> implicit conversion semantics (out of scope of this PR though).